[WIP] Reimplement particles using the Irrlicht particle system

DTA7 opened the pull request:
Addresses #6523, #1414.
I'd expect it not to be slower than the previous implementation for single particles, but faster for particle spawners with many particles, which only use one scene node.
The textures are rotated by 180°, so that they have the correct orientation when rendered, I'm not sure whether this is a bug in Irrlicht.
TODO:
-scale position, velocity and acceleration properly
-add special collision detection for bouncing particles (#2690)
-measure speed
-remove debug outline
-remove vertical property of CE_ADD_PARTICLESPAWNER and CE_SPAWN_PARTICLE

Mod for testing particlespawners: https://pastebin.com/PBKY7FzP

SmallJoker commented on src/particles.cpp:
```diff
@@ -1,23 +1,3 @@
-/*
```
Do not remove the license notice

SmallJoker commented on src/particles.cpp:
```diff
@@ -30,623 +10,697 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-	ISceneNode::OnRegisterSceneNode();
+	return v3f(rand()/(float)RAND_MAX*(max.X-min.X)+min.X,
+		rand()/(float)RAND_MAX*(max.Y-min.Y)+min.Y,
+		rand()/(float)RAND_MAX*(max.Z-min.Z)+min.Z);
```
Needs spaces around all arithmetic operators

SmallJoker commented on src/particles.cpp:
```diff
@@ -30,623 +10,697 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		: last_time(0)
+	{
+		disabled = minacc.X == 0 && minacc.Y == 0 && minacc.Z == 0 &&
+		maxacc.X == 0 && maxacc.Y == 0 && maxacc.Z == 0;
```
Missing indent. Might require a threshold value to compensate calculation errors.

DTA7 replied:
I switched computers, and the indent setting was different...
How large? Like 2*eps or something?

SmallJoker replied:
`minacc.getLengthSQ() < 0.001f && maxacc.getLengthSQ() < 0.001f`, for example.

DTA7 replied:
Would 2e-7 (or a bit larger, in the magnitude of [eps](https://en.wikipedia.org/wiki/Machine_epsilon) for `f32`) be ok too? I think 0.001 is too large.

SmallJoker replied:
I'm not sure how often particles are accelerated by `0.0316 m/s` into one direction. You can use a lower value, such as `1E-6` or even `1E-7`. But I don't have a good feeling when going lower than that to ensure it will be detected as `disabled` when calculation errors sum up.

DTA7 replied:
Ok, I'll make it `1e-6` (an acceleration of 0.001 is still visible)

SmallJoker commented on src/particles.cpp:
```diff
@@ -30,623 +10,697 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+        virtual f32 getRadius() const { return 1.; }
+        virtual irr::scene::E_PARTICLE_EMITTER_TYPE getType() const
+	{
+                return irr::scene::EPET_COUNT;
```
What this class needs is tabulators, not spaces.

Fixer-007 commented:
I've tried homedecor faucets and they did not render :(

DTA7 commented:
@Fixer-007 Thanks, somehow I removed the line that sets the position...

DTA7 commented:
Mod for testing particlespawners: https://pastebin.com/CMkpQA3q

Fixer-007 commented:
I did my test in a fields on actual homedecor faucets, here is my crude results:
![default](https://user-images.githubusercontent.com/16494741/31694065-cdb38d5e-b3aa-11e7-809f-55f91a8f4a4d.png)
I will test with your mod too.

Very nice results.

Fixer-007 commented:
Testing via your mod:
![default](https://user-images.githubusercontent.com/16494741/31721468-25275c4c-b422-11e7-867a-393418038aa2.png)

Fixer-007 commented:
Trying to test it with https://github.com/paramat/snowdrift - simple weather mod, I don't see any particles here for some reason.

DTA7 commented:
It works now, the positioning was quite wrong...

Fixer-007 commented:
I also have suggestion, add particle count in F6 debug info screen, probably easy.

DTA7 commented:
@Fixer-007 With particle spawners seperate?

Fixer-007 commented:
Probably

Fixer-007 commented:
minetest.add_particle gives worse fps (smth like -10/-20%) with this PR :(

Fixer-007 commented:
![default](https://user-images.githubusercontent.com/16494741/31738724-4321da74-b454-11e7-8dd8-a8cd715271e7.png)

DTA7 commented:
@Fixer-007 Like this?
Are `irr::scene::ISceneManager *smgr`,	`ClientEnvironment *env` and `irr::scene::IParticleSystemSceneNode *ps` safe to access (only read) from `handleParticleEvent` and the Irrlicht render thread?

Fixer-007 commented:
No idea here - not a programmer, but devs will probably look at this anyway.

Fixer-007 commented:
Found a bug: when digging in darkness - particles are rendered like it is full day.

DTA7 commented:
@Fixer-007 It was again the position...

numberZero commented:
On my PC, the slowest part is not rendering but collision check.

The reason may be that particles collide with objects and thus use `getObjectsInsideRadius`, which is, er, not too fast... #6453 Is it that important for particles to collide with objects?

Another reason may be that particles are treated as boxes and not as points, that may slow things down as well. That was my first suspect, but it doesn’t seem to be that significant.

DTA7 commented:
@numberZero But if they are treated as points, the collision detection would not have to be as complicated. I think it would be best to add another collision check for particles, which only consideres the node the particle is in, and maybe the one in movement direction, without doing multiple iterations.

paramat commented:
Particles should be treated as points for collision checks i think, or at least make point checking the default and box checking the option. Most particles are small, and perfect collision checking according to their size is not high priority.

Particles colliding with objects is important in some cases, at least we have that option already.

Fixer-007 commented:
Needs rebase. You will continue the work or it is finished?

paramat commented:
@DTA7 any news on this? Does it require adoption by someone else? I feel this is a high priority task.
I have worked with particles a lot in my weather mod and consider the ability to do collision detection essential, so #7519 will not be suitable for weather and many particle uses.
> minetest.add_particle gives worse fps (between 10-20%) with this PR :(, from quick testing on snowdrift mod (it uses add_particle).
update: same fps on low particle count, and worse on anything else.

> Doesn't surprise me. I think it's worth it though, add_particle should be discouraged for anything more than a few particles. Would be nice to see if we could fix this, however.

Needs fixing or a mehtod option if slower. I have found working on snowdrift that using 'add particle' instead of particle spawners is essential in some situations. Particale spawners cannot always replace 'add particle'.
> The reason may be that particles collide with objects [...] Is it that important for particles to collide with objects?

Particles usually do not need to collide with objects, that's a luxury. It should be an option and defualt off.
> Another reason may be that particles are treated as boxes and not as points,

What does the PR do currently? Points or boxes? Points is absolutely fine, boxes can be an option, default off.

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	v3f position;
+};
+
+static video::SColor getParticleLightColor(ClientEnvironment *env, const v3f &position,
```
Why is this global and not a member of `LightingAffector`?

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-	ISceneNode::OnRegisterSceneNode();
+	return v3f(rand() / (float) RAND_MAX * (max.X - min.X) + min.X,
+		rand() / (float) RAND_MAX * (max.Y - min.Y) + min.Y,
+		rand() / (float) RAND_MAX * (max.Z - min.Z) + min.Z);
```
Hmm, I don’t like this distribution. But the API seems to require it.

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+public:
+	LightingAffector(ClientEnvironment *env, IGameDef *gamedef,
+		u8 glow, const video::SColor &base_color = video::SColor(0xFFFFFFFF))
+		:env(env), gamedef(gamedef), glow(glow),
```
Missing space?

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-	));
-}
+		v2f texcoord = anim.getTextureCoords(texsize, animation_frame);
+		texture_matrix = bottomUpTextureMatrix(framesize_ratio.X, framesize_ratio.Y, texcoord.X, texcoord.Y);
```
That’s because texture arrays aren’t supported...

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-	} else {
-		// Spawner exists for an infinity timespan, spawn on a per-second base
+			v3f position = p.pos + camera_offset;
+			v3f velocity = p.vector * 100.f;
```
Why `* 100`?

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		irr::scene::IParticleSystemSceneNode *ps, const video::SColor &color,
+		f32 expirationtime, f32 size, const v3f &vel):
+		smgr(smgr), ps(ps), number(1), deletion_time(0), color(color),
+		expirationtime((u32) expirationtime * 1000.f), size(size), pos(0), vel(vel),
```
`(u32)(expirationtime * 1000.f)`

paramat commented:
Currently inmaster particles do collide with objects, which is intensive, i'll add an option for object collision.

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		ps->addAffector(lighting_affector);
+		lighting_affector->drop();
+
+		ps->setDebugDataVisible(irr::scene::EDS_BBOX);
```
Debug data?

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	case CE_SPAWN_PARTICLE: {
+		scene::IParticleSystemSceneNode *ps =
+			m_smgr->addParticleSystemSceneNode(false,
+				RenderingEngine::get_scene_manager()->getRootSceneNode());
```
I think particles with identical primary properties (texture) might share a particle system. That’s out of scope of this PR, though.

numberZero commented on src/particles.cpp:
```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 {
-	// No particles for "airlike" nodes
-	if (f.drawtype == NDT_AIRLIKE)
-		return;
```
Why removing?

numberZero commented on src/particles.h:
```diff
@@ -181,7 +39,7 @@ friend class ParticleSpawner;
 	~ParticleManager();
 
-	void step (float dtime);
+	//void step (float dtime);
```
A sign of WIP.

numberZero commented:
> however since this is the standard for collision detection there may be no benefit to using a point, as that would just be reducing the box to a very small size.

There will be a benefit, will that be implemented as a separate function knowing that’s just a point. Using a tiny box instead is virtually pointless ofc.

paramat commented:
No response from author or progress, will re-open on progress, or needs adoption.

DTA7 commented:
I've included your suggestions and implemented the special collision handling for points, as well as addressed the previous TODOs.
-The scaling is correct now, the velocities for the Irrlicht particles are given per millisecond. @numberZero
-The collision detection is much faster, about ~5 times when I tested it (by doing the calculation
with both methods by uncommenting the `collisionMoveSimple` code in "particles.cpp" of 46dc879cf143fb838b6db3c9e330974762dbe0fe and comparing the Profiler info). This speedup comes from the much lower amount of nodes considered by the treatment as points.
-I added the `bounce_fraction` and the `bounce_threshold` properties to specify the retained fraction of the velocity component in the direction of the bounce, and the threshold for the ratio of this component to the total speed, below which the particle stops completely. This causes the "digging particles" not to slide anymore as before.
-The `vertical` property was removed.
-To prevent particles partially hidden in nodes (which also causes wrong lighting), the colliding nodeboxes are treated as enlarged by the particle size. This causes problems on the edges of nodes for large particles, but to change this more nodes have to be considered. For the most part, this is not really noticable.

The changed properties require a change in the network protocol, so the change in 9c57ce0122b54fc26f57fd8ab77d7de5e972ec73 could be removed in this pull request.

[Updated mod for testing](https://pastebin.com/sBKE13jv)

@paramat Could you reopen the pull request?

paramat commented:
Github won't let me because "the branch was force-pushed or recreated". Seems a new PR is needed.
